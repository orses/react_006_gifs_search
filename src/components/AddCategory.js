import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState('');

  const handleInputChange = e => {
    setInputValue(e.target.value);
  };

  const handleSubmit = e => {
    e.preventDefault();
    // updating categories state
    if (inputValue.trim().length > 2) {
      setCategories(categories => [inputValue, ...categories]);
      setInputValue('');
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <input
        type="search"
        value={inputValue}
        onChange={handleInputChange}
        placeholder="Type a category and hit Enter"
      />
    </form>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
