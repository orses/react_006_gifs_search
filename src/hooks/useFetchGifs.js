import { useState, useEffect } from 'react';
import getGifs from '../helpers/getGifs';

export const useFetchGifs = category => {
  const [state, setState] = useState({
    data: [],
    loading: true,
  });

  // se ejecuta al montarse, pero no al renderizarse dado que tiene una dependencia declarada y se le ha pasado category, que en este componente nunca cambia
  useEffect(() => {
    getGifs(category).then(images =>
      setState({
        data: images,
        loading: false,
      })
    );
  }, [category]);

  return state;
};
