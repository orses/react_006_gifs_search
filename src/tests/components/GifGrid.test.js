import React from 'react';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('test on GifGrid component', () => {
  const category = 'cat';
  let wrapper;

  test('should render the component', () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });
    wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should show items when useFetchGifs to load images', () => {
    const gifs = [
      {
        id: 'abc',
        url: 'http://localhost:4000/cualquiercosa.jpg',
        title: 'Una imagen imaginaria',
      },
      {
        id: 'eca',
        url: 'http://localhost:4000/cualquierotracosa.jpg',
        title: 'Una imagen imaginaria m√°s',
      },
    ];

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });

    wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('p').exists()).toBe(false); // no loading paragraph
    expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
  });
});
