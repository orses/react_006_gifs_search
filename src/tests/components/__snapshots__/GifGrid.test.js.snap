// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test on GifGrid component should render the component 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    cat
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`test on GifGrid component should show items when useFetchGifs to load images 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    cat
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="abc"
      key="abc"
      title="Una imagen imaginaria"
      url="http://localhost:4000/cualquiercosa.jpg"
    />
    <GifGridItem
      id="eca"
      key="eca"
      title="Una imagen imaginaria mÃ¡s"
      url="http://localhost:4000/cualquierotracosa.jpg"
    />
  </div>
</Fragment>
`;
